grudge/geometry/metrics.py

```diff
@@ -0,0 +1,713 @@
+            grp_field[i] = \
+                (2.0 * (batch.to_element_face % 2) - 1.0) * grp_field[i]
+
+    return signed_face_ones
+
+
+def parametrization_derivative(
+        actx: ArrayContext, dcoll: DiscretizationCollection, dd
+        ) -> MultiVector:
+    r"""Computes the product of forward metric derivatives spanning the
+    tangent space with topological dimension *dim*.
+
+    :arg dd: a :class:`~grudge.dof_desc.DOFDesc`, or a value convertible to one.
+        Defaults to the base volume discretization.
+    :returns: a :class:`pymbolic.geometric_algebra.MultiVector` containing
+        the product of metric derivatives.
+    """
+    if dd is None:
+        dd = DD_VOLUME
+
+    dim = dcoll.discr_from_dd(dd).dim
+    if dim == 0:
+        from pymbolic.geometric_algebra import get_euclidean_space
+
+        return MultiVector(
+            _signed_face_ones(actx, dcoll, dd),
+            space=get_euclidean_space(dcoll.ambient_dim)
+        )
+
+    from pytools import product
+
+    return product(
+        forward_metric_derivative_mv(actx, dcoll, rst_axis, dd)
+        for rst_axis in range(dim)
+    )
+
+
+def pseudoscalar(actx: ArrayContext, dcoll: DiscretizationCollection,
+        dd=None) -> MultiVector:
+    r"""Computes the field of pseudoscalars for the domain/discretization
+    identified by *dd*.
+
+    :arg dd: a :class:`~grudge.dof_desc.DOFDesc`, or a value convertible to one.
+        Defaults to the base volume discretization.
+    :returns: A :class:`~pymbolic.geometric_algebra.MultiVector` of
+        :class:`~meshmode.dof_array.DOFArray`\ s.
+    """
+    if dd is None:
+        dd = DD_VOLUME
+
+    return parametrization_derivative(actx, dcoll, dd).project_max_grade()
+
+
+def area_element(
+        actx: ArrayContext, dcoll: DiscretizationCollection, dd=None
+        ) -> DOFArray:
+    r"""Computes the scale factor used to transform integrals from reference
+    to global space.
+
+    :arg dd: a :class:`~grudge.dof_desc.DOFDesc`, or a value convertible to one.
+        Defaults to the base volume discretization.
+    :returns: a :class:`~meshmode.dof_array.DOFArray` containing the transformed
+        volumes for each element.
+    """
+    if dd is None:
+        dd = DD_VOLUME
+
+    dim = dcoll.discr_from_dd(dd).dim
+
+    @memoize_in(dcoll, (area_element, dd,
+                        "area_elements_adim%s_gdim%s"
+                        % (dcoll.ambient_dim, dim)))
+    def _area_elements():
+        return actx.np.sqrt(
+            pseudoscalar(actx, dcoll, dd=dd).norm_squared()
+        )
+    return _area_elements()
+
+# }}}
+
+
+# {{{ Surface normal vectors
+
+def rel_mv_normal(
+        actx: ArrayContext, dcoll: DiscretizationCollection, dd=None) -> MultiVector:
+    r"""Computes surface normals at each nodal location as a
+    :class:`~pymbolic.geometric_algebra.MultiVector` relative to the 'volume'
+    discretization's surface pseudoscalar.
```

Comment:
```suggestion
    :class:`~pymbolic.geometric_algebra.MultiVector` relative to the
    surface pseudoscalar of the discretization described by *dd*.
```
? `'volume'` sounds a bit weird to me because this tends to compute normals on faces.

Reply:
It does. #103 for the bigger issue.

Reply:
I'm fine with just leaving this as is, since my suggestion isn't particularly clearer.